File: backend/app/services/gemini_file_service.py
import os
import tempfile
import logging
import httpx
from typing import Optional
from google import genai
from google.genai import types
from app.config import settings
from app.services.supabase_storage import download_pdf_from_supabase
from app.agents.gemini_client import get_client as get_gemini_client

logger = logging.getLogger("quizloop.gemini_file")

async def upload_file_to_gemini(file_path: str, mime_type: str = "application/pdf") -> str:
    """
    If using Gemini Developer client (API Key), uploads to Gemini File API and returns URI.
    If using Vertex AI, returns file_path directly.
    """
    client = get_gemini_client()
    if not settings.google_cloud_project:
        try:
            logger.info(f"Uploading file to Gemini File API: {file_path}")
            file_ref = client.files.upload(
                file=file_path,
                config=dict(mime_type=mime_type)
            )
            logger.info(f"Successfully uploaded to Gemini File API: {file_ref.name} ({file_ref.uri})")
            return file_ref.uri
        except Exception as e:
            logger.warning(f"Gemini File API upload skipped/failed: {e}. Falling back to byte transmission.")
    
    return file_path

async def get_gemini_part_for_file(file_uri_or_path: str, mime_type: str = "application/pdf") -> types.Part:
    """
    Unified resolver: converts a Gemini File URI, HTTP URL (Supabase), or local file path
    into a valid google.genai.types.Part for generate_content.
    """
    if file_uri_or_path.startswith("https://generativelanguage.googleapis.com"):
        return types.Part.from_uri(file_uri=file_uri_or_path, mime_type=mime_type)
    
    if os.path.exists(file_uri_or_path):
        with open(file_uri_or_path, "rb") as f:
            data = f.read()
        return types.Part.from_bytes(data=data, mime_type=mime_type)
    
    if file_uri_or_path.startswith("http://") or file_uri_or_path.startswith("https://"):
        async with httpx.AsyncClient() as client:
            resp = await client.get(file_uri_or_path, follow_redirects=True)
            resp.raise_for_status()
            data = resp.content
        return types.Part.from_bytes(data=data, mime_type=mime_type)
    
    # Try downloading from Supabase storage by filename
    try:
        data = await download_pdf_from_supabase(file_uri_or_path)
        return types.Part.from_bytes(data=data, mime_type=mime_type)
    except Exception as e:
        raise ValueError(f"Could not resolve file part from '{file_uri_or_path}': {e}")

async def ensure_valid_gemini_file(file_uri: Optional[str], file_name: Optional[str]) -> str:
    """
    Checks if a Gemini file URI is still valid. If expired (48h TTL) or missing,
    downloads from Supabase Storage and re-uploads to Gemini File API.
    """
    client = get_gemini_client()
    
    if file_uri and file_uri.startswith("https://generativelanguage.googleapis.com"):
        try:
            file_name_id = file_uri.split("/")[-1]
            file_info = client.files.get(name=f"files/{file_name_id}")
            if file_info.state.name == "ACTIVE":
                return file_uri
        except Exception as e:
            logger.warning(f"Gemini file URI {file_uri} expired or not found: {e}. Re-uploading...")

    if not file_name:
        raise ValueError("Cannot re-upload file: file_name is missing.")

    pdf_bytes = await download_pdf_from_supabase(file_name)
    with tempfile.NamedTemporaryFile(suffix=".pdf", delete=False) as tmp:
        tmp.write(pdf_bytes)
        tmp_path = tmp.name

    try:
        new_uri = await upload_file_to_gemini(tmp_path, "application/pdf")
        return new_uri
    finally:
        if os.path.exists(tmp_path):
            os.remove(tmp_path)

File: src/server/services/gemini-file-service.ts
import fs from "fs";
import os from "os";
import path from "path";
import crypto from "crypto";
import { getServerConfig } from "../config";
import { downloadPdfFromSupabase } from "./supabase-storage";
import { getGeminiClient } from "../agents/gemini-client";

export async function uploadFileToGemini(
  filePathOrBuffer: string | Buffer,
  mimeType = "application/pdf"
): Promise<string> {
  const config = getServerConfig();

  // If using Google Cloud Vertex AI without Developer API key, return the file path directly
  if (config.googleCloudProject && !config.geminiApiKey) {
    if (typeof filePathOrBuffer === "string") {
      return filePathOrBuffer;
    }
    // Write buffer to temp file
    const tmpPath = path.join(os.tmpdir(), `gemini_upload_${Date.now()}_${crypto.randomBytes(4).toString("hex")}.pdf`);
    fs.writeFileSync(tmpPath, filePathOrBuffer);
    return tmpPath;
  }

  try {
    const client = getGeminiClient();
    let uploadTarget: string | Blob;

    if (typeof filePathOrBuffer === "string") {
      uploadTarget = filePathOrBuffer;
    } else {
      uploadTarget = new Blob([new Uint8Array(filePathOrBuffer)], { type: mimeType });
    }

    const fileRef = await client.files.upload({
      file: uploadTarget as any,
      config: { mimeType },
    });

    if (fileRef && fileRef.uri) {
      return fileRef.uri;
    }
  } catch (err) {
    console.warn("Gemini File API upload skipped/failed:", err, "Falling back to byte transmission.");
  }

  if (typeof filePathOrBuffer === "string") {
    return filePathOrBuffer;
  }

  const tmpPath = path.join(os.tmpdir(), `gemini_fallback_${Date.now()}_${crypto.randomBytes(4).toString("hex")}.pdf`);
  fs.writeFileSync(tmpPath, filePathOrBuffer);
  return tmpPath;
}

export interface GeminiFilePart {
  fileData?: {
    fileUri: string;
    mimeType: string;
  };
  inlineData?: {
    data: string;
    mimeType: string;
  };
}

export async function getGeminiPartForFile(
  fileUriOrPath: string,
  mimeType = "application/pdf"
): Promise<GeminiFilePart> {
  // 1. Gemini File API URI
  if (fileUriOrPath.startsWith("https://generativelanguage.googleapis.com")) {
    return {
      fileData: {
        fileUri: fileUriOrPath,
        mimeType,
      },
    };
  }

  // 2. Local filesystem path
  if (fs.existsSync(fileUriOrPath)) {
    const buffer = fs.readFileSync(fileUriOrPath);
    return {
      inlineData: {
        data: buffer.toString("base64"),
        mimeType,
      },
    };
  }

  // 3. HTTP(S) URL
  if (fileUriOrPath.startsWith("http://") || fileUriOrPath.startsWith("https://")) {
    try {
      const resp = await fetch(fileUriOrPath);
      if (!resp.ok) {
        throw new Error(`HTTP status ${resp.status}`);
      }
      const arrayBuf = await resp.arrayBuffer();
      return {
        inlineData: {
          data: Buffer.from(arrayBuf).toString("base64"),
          mimeType,
        },
      };
    } catch (e) {
      console.warn(`Direct URL fetch failed for ${fileUriOrPath}:`, e);
    }
  }

  // 4. Supabase Storage fallback
  try {
    const buffer = await downloadPdfFromSupabase(fileUriOrPath);
    return {
      inlineData: {
        data: buffer.toString("base64"),
        mimeType,
      },
    };
  } catch (e) {
    throw new Error(`Could not resolve file part from '${fileUriOrPath}': ${e}`);
  }
}

export async function ensureValidGeminiFile(
  fileUri: string | null,
  fileName: string | null
): Promise<string> {
  const client = getGeminiClient();

  if (fileUri && fileUri.startsWith("https://generativelanguage.googleapis.com")) {
    try {
      const parts = fileUri.split("/");
      const fileId = parts[parts.length - 1];
      const fileInfo = await client.files.get({ name: `files/${fileId}` });
      if (fileInfo && (fileInfo as any).state === "ACTIVE") {
        return fileUri;
      }
    } catch (err) {
      console.warn(`Gemini file URI ${fileUri} expired or not found:`, err, "Re-uploading...");
    }
  }

  if (!fileName) {
    throw new Error("Cannot re-upload file: fileName is missing.");
  }

  const pdfBytes = await downloadPdfFromSupabase(fileName);
  const tmpPath = path.join(os.tmpdir(), `gemini_reupload_${Date.now()}_${crypto.randomBytes(4).toString("hex")}.pdf`);
  fs.writeFileSync(tmpPath, pdfBytes);

  try {
    const newUri = await uploadFileToGemini(tmpPath, "application/pdf");
    return newUri;
  } finally {
    if (fs.existsSync(tmpPath)) {
      try {
        fs.unlinkSync(tmpPath);
      } catch {
        // ignore
      }
    }
  }
}
